From 085327159973a2c4b98c23548c08937759e29ccb Mon Sep 17 00:00:00 2001
From: wu xia <xia.wu@intel.com>
Date: Tue, 12 Mar 2013 16:01:52 +0800
Subject: [PORT FROM MAIN] [PATCH] sqlite: enable ATOMIC_WRITE feature for sqlite lib

BZ: 90764

ATOMIC_WRITE feature will improve SQLite performance, especially for
small updates(data size is less than one page (4KB)).

Change-Id: Ifddf59fc83b6a43d7fc3367b6492b0608f97325c
Orig-Change-Id: Ia69099af89402439d2466c9aa4102077ebb815f1
Signed-off-by: wu xia <xia.wu@intel.com>
Reviewed-on: http://android.intel.com:8080/95075
Reviewed-by: cactus <cactus@intel.com>
Tested-by: Sun, Jianhua <jianhua.sun@intel.com>
Reviewed-by: Huang, Bo B <bo.b.huang@intel.com>
Reviewed-by: Tang, Richard <richard.tang@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
---
 dist/Android.mk |    1 +
 dist/sqlite3.c  |   14 +++++++++++---
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/dist/Android.mk b/dist/Android.mk
index 0ad8e44..4f32371 100644
--- a/dist/Android.mk
+++ b/dist/Android.mk
@@ -24,6 +24,7 @@ common_sqlite_flags := \
 	-DSQLITE_ENABLE_FTS3 \
 	-DSQLITE_ENABLE_FTS3_BACKWARDS \
 	-DSQLITE_ENABLE_FTS4 \
+	-DSQLITE_ENABLE_ATOMIC_WRITE \
 	-DSQLITE_OMIT_BUILTIN_TEST \
 	-DSQLITE_OMIT_COMPILEOPTION_DIAGS \
 	-DSQLITE_OMIT_LOAD_EXTENSION \
diff --git a/dist/sqlite3.c b/dist/sqlite3.c
index a43031b..92caab7 100644
--- a/dist/sqlite3.c
+++ b/dist/sqlite3.c
@@ -28673,11 +28673,19 @@ static int unixSectorSize(sqlite3_file *pFile){
 */
 static int unixDeviceCharacteristics(sqlite3_file *id){
   unixFile *p = (unixFile*)id;
+  int returnValue = 0;
+
   if( p->ctrlFlags & UNIXFILE_PSOW ){
-    return SQLITE_IOCAP_POWERSAFE_OVERWRITE;
-  }else{
-    return 0;
+    returnValue = SQLITE_IOCAP_POWERSAFE_OVERWRITE;
   }
+  // Begin Android change
+  // Enable ATOMIC_WRITE feature by default
+  #ifdef SQLITE_ENABLE_ATOMIC_WRITE
+  returnValue |= SQLITE_IOCAP_ATOMIC4K;
+  #endif
+  // End Android change
+
+  return returnValue;
 }
 
 #ifndef SQLITE_OMIT_WAL
-- 
1.7.4.1

